File: sdks/node/tests/errors.test.ts
import { describe, it, expect } from "vitest";
import {
  DocDigitizerError,
  AuthenticationError,
  ValidationError,
  RateLimitError,
  ServerError,
  ServiceUnavailableError,
  TimeoutError,
  raiseForStatus,
} from "../src/errors.js";

describe("Error hierarchy", () => {
  it("all errors extend DocDigitizerError", () => {
    const errors = [
      new AuthenticationError("auth"),
      new ValidationError("val"),
      new RateLimitError("rate"),
      new ServerError("srv"),
      new ServiceUnavailableError("unavail"),
      new TimeoutError("timeout"),
    ];
    for (const err of errors) {
      expect(err).toBeInstanceOf(DocDigitizerError);
      expect(err).toBeInstanceOf(Error);
    }
  });

  it("carries statusCode, traceId, messages, timers", () => {
    const err = new DocDigitizerError(
      "test",
      400,
      "TRACE1",
      ["msg1", "msg2"],
      { total: 123 }
    );
    expect(err.message).toBe("test");
    expect(err.statusCode).toBe(400);
    expect(err.traceId).toBe("TRACE1");
    expect(err.messages).toEqual(["msg1", "msg2"]);
    expect(err.timers).toEqual({ total: 123 });
  });

  it("has correct name for each error class", () => {
    expect(new AuthenticationError("x").name).toBe("AuthenticationError");
    expect(new ValidationError("x").name).toBe("ValidationError");
    expect(new RateLimitError("x").name).toBe("RateLimitError");
    expect(new ServerError("x").name).toBe("ServerError");
    expect(new ServiceUnavailableError("x").name).toBe("ServiceUnavailableError");
    expect(new TimeoutError("x").name).toBe("TimeoutError");
  });
});

describe("raiseForStatus", () => {
  it("throws ValidationError for 400", () => {
    expect(() => raiseForStatus(400, "T1", ["bad request"])).toThrow(
      ValidationError
    );
  });

  it("throws AuthenticationError for 401", () => {
    expect(() => raiseForStatus(401)).toThrow(AuthenticationError);
  });

  it("throws RateLimitError for 429", () => {
    expect(() => raiseForStatus(429)).toThrow(RateLimitError);
  });

  it("throws ServerError for 500", () => {
    expect(() => raiseForStatus(500)).toThrow(ServerError);
  });

  it("throws ServiceUnavailableError for 503", () => {
    expect(() => raiseForStatus(503)).toThrow(ServiceUnavailableError);
  });

  it("throws TimeoutError for 504", () => {
    expect(() => raiseForStatus(504)).toThrow(TimeoutError);
  });

  it("throws DocDigitizerError for unknown status codes", () => {
    expect(() => raiseForStatus(418)).toThrow(DocDigitizerError);
  });

  it("joins messages into the error message", () => {
    try {
      raiseForStatus(400, undefined, ["err1", "err2"]);
    } catch (e) {
      expect((e as Error).message).toBe("err1; err2");
    }
  });

  it("falls back to HTTP status when no messages", () => {
    try {
      raiseForStatus(500);
    } catch (e) {
      expect((e as Error).message).toBe("HTTP 500");
    }
  });
});

File: sdks/rust/tests/errors_test.rs
use docdigitizer::errors::{raise_for_status, is_retryable, DocDigitizerError};

#[test]
fn test_raise_for_status_400() {
    let err = raise_for_status(400, Some("T1".into()), Some(vec!["bad request".into()]), None);
    assert!(matches!(err, DocDigitizerError::Validation { .. }));
    assert_eq!(err.to_string(), "bad request");
}

#[test]
fn test_raise_for_status_401() {
    let err = raise_for_status(401, None, None, None);
    assert!(matches!(err, DocDigitizerError::Authentication { .. }));
    assert_eq!(err.to_string(), "HTTP 401");
}

#[test]
fn test_raise_for_status_429() {
    let err = raise_for_status(429, None, None, None);
    assert!(matches!(err, DocDigitizerError::RateLimit { .. }));
}

#[test]
fn test_raise_for_status_500() {
    let err = raise_for_status(500, None, None, None);
    assert!(matches!(err, DocDigitizerError::Server { .. }));
}

#[test]
fn test_raise_for_status_503() {
    let err = raise_for_status(503, None, None, None);
    assert!(matches!(err, DocDigitizerError::ServiceUnavailable { .. }));
}

#[test]
fn test_raise_for_status_504() {
    let err = raise_for_status(504, None, None, None);
    assert!(matches!(err, DocDigitizerError::Timeout { .. }));
}

#[test]
fn test_raise_for_status_unknown() {
    let err = raise_for_status(418, None, None, None);
    assert!(matches!(err, DocDigitizerError::Unknown { .. }));
}

#[test]
fn test_raise_for_status_joins_messages() {
    let err = raise_for_status(400, None, Some(vec!["err1".into(), "err2".into()]), None);
    assert_eq!(err.to_string(), "err1; err2");
}

#[test]
fn test_is_retryable() {
    assert!(is_retryable(429));
    assert!(is_retryable(500));
    assert!(is_retryable(503));
    assert!(is_retryable(504));
    assert!(!is_retryable(400));
    assert!(!is_retryable(401));
    assert!(!is_retryable(404));
}

#[test]
fn test_error_carries_fields() {
    let err = raise_for_status(401, Some("TRACE1".into()), Some(vec!["bad key".into()]), None);
    match err {
        DocDigitizerError::Authentication { trace_id, messages, .. } => {
            assert_eq!(trace_id, Some("TRACE1".to_string()));
            assert_eq!(messages, Some(vec!["bad key".to_string()]));
        }
        _ => panic!("expected Authentication error"),
    }
}
